wordpress-mobile/WordPress-iOS
# [Tooling] Add Buildkite annotation on test failures

This PR uses [Buildkite annotations](https://buildkite.com/docs/agent/v3/cli-annotate) to highlight the Unit and UI Tests that might have failed during a CI build.

![Example](https://user-images.githubusercontent.com/216089/176279184-19ce5eb4-9af5-438f-ab65-026d638008c5.png)

## To Test

 - If you see the build either fail on UI or Unit Tests because of flaky tests and those test failures being reported as annotations, that's great
 - If not and all test pass as they should (even the flaky ones), it's also a great sign that successful tests don't break (or try to annotate with an empty text, for example)
 - You can then use https://github.com/wordpress-mobile/WordPress-iOS/pull/18955 to test this PR by introducing test failure. In its current state, it not only introduces intentional fake failures in the tests, but also tweaks the pipeline to make it stub the junit file (useful to iterate faster when debugging this stuff!), but feel free to play with it to test more thoroughly.

## Alternatives Considered

This PR uses the route of using a simple XSLT to extract the failure cases from the JUnit report in a concise markdown text, which keeps the pipeline simpler.

I also tried implementing the same thing but using Buildkite's `junit-annotate` plugin in https://github.com/wordpress-mobile/WordPress-iOS/pull/18953, to compare the two approaches. After experimenting with that alternate approach, I realized that it created too much noise not only in the `pipeline.yml`, but also in the Build page (where the `wait` and `plugins: junit-annotate` additional steps are a bit "too" visible in the list of steps), and lastly required waiting for a new agent and spawning a docker container all that just for creating the annotation.

This is why in the end I found that the solution in this PR #18952 was not only simpler but also better and less resource-consuming.

## (ex-)Related PR

While working on this and discovering Buildkite's ability to group steps, I also played with it a bit to restructure the order of the steps in the pipeline. Those pipeline changes were initially built on top of this PR (hence outdated comments below), but I've since extracted them [in a separate PR #18954](https://github.com/wordpress-mobile/WordPress-iOS/pull/18954) (based on `trunk`) to isolate each unrelated changes in their own PR.
